torch/_subclasses/fake_tensor.py

```diff
@@ -0,0 +1,140 @@
+import torch
+
+from torch._subclasses import BaseTensor
+from torch.utils._pytree import tree_map
+from functools import partial
+from torch.fx.operator_schemas import normalize_function
+from typing import Union
+
+aten = torch.ops.aten
+
+_device_not_kwarg_ops = (
+    aten._resize_output_.default,
+    aten.nested_tensor.default,
+    aten.pin_memory.default,
+    aten.is_pinned.default,
+    aten.to.device,
+    aten.to.prim_Device,
+    aten._pin_memory.default,
+)
+
+# Meta tensors give you the ability to run PyTorch code without having to
+# actually do computation through tensors allocated on a `meta` device.
+# Because the device is `meta`, meta tensors do not model device propagation.
+# FakeTensor extends MetaTensors to also carry an additional `fake_device`
+# which tracks devices that would have been used.
+
+
+class FakeTensor(BaseTensor):
+    fake_device: torch.device
+
+    @staticmethod
+    def __new__(cls, elem, device):
+        return super().__new__(cls, elem)
+
+    def __init__(self, elem, device: Union[torch.device, str]):
+        # elem does not need to be recorded, because FakeTensor *is a* elem
+        assert elem.device.type == "meta"
+        device = device if isinstance(device, torch.device) else torch.device(device)
+        assert device.type != "meta"
+        self.fake_device = device
+
+    @staticmethod
+    def from_tensor(t):
+        existing_device = t.device
+        return FakeTensor(t.to(device="meta"), existing_device)
```

Comment:
It probably doesn't matter for most users, but maybe it is necessary to do something fancier https://github.com/pytorch/pytorch/blob/50cadfae10fe61b6daa0d4f39a3a97fa19bc82eb/test/test_meta.py#L70-L203

Reply:
Resolved in follow up pr https://github.com/pytorch/pytorch/pull/78090